include/convex_bodies/ellipsoid.h

```diff
@@ -0,0 +1,307 @@
+                A(i,j) = Ain[i][j];
+            }
+        }
+
+        Eigen::SelfAdjointEigenSolver<MT> eigensolver(A);
+        if (eigensolver.info() != Eigen::Success) {
+            throw std::runtime_error("Eigen solver returned error!");
+        }
+
+        eigen_values = eigensolver.eigenvalues();
+        eigen_vecs = eigensolver.eigenvectors();
+
+        eigen_values_inv = eigen_values.array().inverse().matrix();
+        eigen_values_inv_sqrt = eigen_values_inv.array().sqrt().matrix();
+
+        dim = A.rows();
+        c = Point(dim);
+        c.set_to_origin();
+    }
+
+
+    VT eigenvals() const {
+        return eigen_values;
+    }
+
+
+    VT eigenvals_inv() const {
+        return eigen_values_inv;
+    }
+
+
+    VT eigenvals_inv_sqrt() const {
+        return eigen_values_inv_sqrt;
+    }
+
+
+    MT eigenvecs() const {
+        return eigen_vecs;
+    }
+
+
+    unsigned int dimensions() const {
+        return dim;
+    }
+
+
+    void print() {
+        std::cout << "Ellipse is in the form: (x-c)' A (x-c) <= 1 \n";
+        std::cout << "c = \n" << c.print();
+        std::cout << "A = \n" << A;
+    }
+
+
+
+    NT mat_mult(Point const& p) {
+        VT x = p.getCoefficients();
+
+        if (dim < 15) {
+            NT sum = 0;
+            for (Eigen::Index i = 0; i < dim; ++i) {
+                const auto x_i = x[i];
+                sum += A(i, i) * x_i * x_i;
+                for (Eigen::Index j = 0; j < i; ++j) {
+                    sum += 2 * A(j, i) * x_i * x[j];
+                }
+            }
+            return sum;
+        }
+        
+        return x.transpose() * A.template selfadjointView<Eigen::Upper>() * x;
+    }
+
+    
+    NT log_volume () {
+        NT ball_log_vol = (NT(dim)/NT(2) * std::log(M_PI)) - log_gamma_function(NT(dim) / NT(2) + 1);
+        NT det_factor = - 0.5 * std::log( A.determinant() );
+
+        return det_factor + ball_log_vol;
+    }
+
+
+    void scale(NT scale_factor) {
+        assert (scale_factor > 0);
+
+        NT scale_factor_sq = scale_factor * scale_factor;
+        NT inv_scale_factor = (NT(1.0) / scale_factor);
+        NT inv_scale_factor_sq = (NT(1.0) / scale_factor_sq);
+
+        // L = mult_factor * L;
+        eigen_values = inv_scale_factor_sq * eigen_values;
+        eigen_values_inv = scale_factor_sq * eigen_values_inv;
+        eigen_values_inv_sqrt = scale_factor * eigen_values_inv_sqrt;
+
+        A = inv_scale_factor_sq * A; // as volume depends on square root of it's determinant
+    }
+
+
+    int is_in(Point const& p){
+        NT val = mat_mult(p - c);
+        if (val > 1) {
+            return 0;
+        }
+
+        return -1;
+    }
+
+
+    // compute intersection point of ray starting from r and pointing to v
+    std::pair<NT, NT> line_intersect(Point& r, Point& v) const {
+        // constants of a quadratic equation
+        NT a_q = mat_mult(v);
+        NT b_q = 2 * (r - c).getCoefficients().dot(A * v.getCoefficients());
```

Comment:
A is symmetric right? Could you use the Eigen's optimization here for the matrix-vector multiplication?

Reply:
Added a `vec_mult` function,